File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
from collections import deque

def solution(maps):
    answer = []
    n = len(maps)
    m = len(maps[0])
    visit = [[False]*m for _ in range(n)]
    for i in range(n):
        for j in range(m):
            if visit[i][j] == False and maps[i][j] != 'X':
                answer.append(bfs(maps, visit,n,m,i,j))
    if not answer:
        return [-1]
    answer.sort()
    return answer
    
def bfs(maps,visit,n,m,i,j):
    dx = [-1,1,0,0]
    dy = [0,0,-1,1]
    que = deque()
    que.append((i,j))
    visit[i][j] = True
    days=0
    while que:
        x,y = que.popleft()
        days+=int(maps[x][y])
        for i in range(4):
            nx = dx[i]+x
            ny = dy[i]+y
            if 0<=nx<n and 0<=ny<m and visit[nx][ny] == False and maps[nx][ny] !='X':
                que.append((nx,ny))
                visit[nx][ny]=True
    return days
    

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.java
import java.util.*;

class Solution {
    
    static boolean[][] visit;
    
    static List<Integer> arr = new ArrayList<>();
    static int[] dx = {-1,0,1,0};
    static int[] dy = {0,1,0,-1};
    
    
    public int[] solution(String[] maps) {
        int[] answer;
        visit = new boolean[maps.length][maps[0].length()];
        
        for(int i=0;i<maps.length;i++){
            for(int j=0;j<maps[0].length();j++){
                if(!visit[i][j] && maps[i].charAt(j) !='X'){
                    arr.add(bfs(maps, i, j));
                }
            }
        }
        if(arr.size() == 0) return new int[]{-1};
        Collections.sort(arr);
        answer= new int[arr.size()];
        for(int i=0;i<arr.size();i++)answer[i] = arr.get(i);
        
        return answer;
    }
    
    public int bfs(String[] maps, int i, int j){
        
        Queue<int[]> que = new LinkedList<>();
        que.add(new int[]{i,j});
        visit[i][j] = true;
        int cnt=0;
        
        while(!que.isEmpty()){
            int[] cur = que.poll();
            cnt += maps[cur[0]].charAt(cur[1])-'0';
            for(int d=0;d<4;d++){
                int nx = cur[1]+dx[d];
                int ny = cur[0]+dy[d];
                if(nx<0 || nx>=maps[0].length() || ny<0 || ny>=maps.length) continue;
                if(maps[ny].charAt(nx) != 'X' && !visit[ny][nx]){
                    visit[ny][nx] = true;
                    que.add(new int[]{ny, nx});
                }
            }
        }
        return cnt;
    }
    
}